File: Parser.java
//
// Created by ozrenkosi on 25.04.2017..
//

import java.util.*;

public class Parser {

    private static List<String> inputString = new ArrayList<>();
    private static boolean parserOK;

    public static void main(String[] args) {

        dataInput();

        parseString_S();

        checkDefinition();

    }

    private static void dataInput() {
        String[] temporaryInput;

        Scanner reader = new Scanner(System.in);

        temporaryInput = reader.nextLine().split("");

        inputString.addAll(Arrays.asList(temporaryInput));

        reader.close();
    }

    private static void parseString_S() {
        parserOK = false;
        System.out.print("S");

        if (!inputString.isEmpty() && inputString.get(0).equals("b")) {
            inputString.remove(0);
            parseString_B();
            if (parserOK) {
                parseString_A();
            }
        }
        else if (!inputString.isEmpty() && inputString.get(0).equals("a")) {
            inputString.remove(0);
            parseString_A();
            if (parserOK) {
                parseString_B();
            }
        }
        else {
            parserOK = false;
        }
    }

    private static void parseString_A() {
        parserOK = false;
        System.out.print("A");

        if (!inputString.isEmpty() && inputString.get(0).equals("b")) {
            inputString.remove(0);
            parseString_C();
        }
        else if (!inputString.isEmpty() && inputString.get(0).equals("a")) {
            inputString.remove(0);
            parserOK = true;
        }
        else {
            parserOK = false;
        }
    }

    private static void parseString_B() {
        parserOK = false;
        System.out.print("B");

        if (inputString.isEmpty()) {
            parserOK = true;
        }
        else {
            parserOK = false;
            if (inputString.get(0).equals("c")) {
                inputString.remove(0);
                if (!inputString.isEmpty() && inputString.get(0).equals("c")) {
                    inputString.remove(0);
                    parseString_S();
                    if (parserOK && !inputString.isEmpty() && inputString.get(0).equals("b")) {
                        inputString.remove(0);
                        if (!inputString.isEmpty() && inputString.get(0).equals("c")) {
                            inputString.remove(0);
                            parserOK = true;
                        }
                        else {
                            parserOK = false;
                        }
                    }
                    else {
                        parserOK = false;
                    }
                }
                else {
                    parserOK = false;
                }
            }
            else {
                parserOK = true;
            }
        }
    }

    private static void parseString_C() {
        parserOK = false;
        System.out.print("C");

        parseString_A();
        if (parserOK) {
            parseString_A();
        }
    }

    private static void checkDefinition() {
        System.out.println();

        if (parserOK && inputString.isEmpty()) {
            System.out.println("DA");
        }
        else {
            System.out.println("NE");
        }
    }

}

File: Parser.swift
//
//  Parser.swift
//  Parser
//
//  Created by Ozren Kosi on 04.05.2017..
//  Copyright © 2017. Ozren Kosi. All rights reserved.
//

import Foundation

var inputString = [String]()
var parserOK = Bool()

func dataInput() {
    let temporaryInput = readLine()!
    inputString = temporaryInput.characters.map { String($0) }
}

func parseString_S() {
    parserOK = false
    print("S", terminator:"")
    
    if (inputString.isEmpty) {
        parserOK = false
        return
    }
    else if (inputString[0] == "b") {
        inputString.remove(at: 0)
        parseString_B()
        if (parserOK) {
            parseString_A()
        }
        return
    }
    
    if (inputString.isEmpty) {
        parserOK = false
    }
    else if (inputString[0] == "a") {
        inputString.remove(at: 0)
        parseString_A()
        if (parserOK) {
            parseString_B()
            
        }
    }
}

func parseString_A() {
    parserOK = false
    print("A", terminator:"")
    
    if (inputString.isEmpty) {
        parserOK = false
        return
    }
    else if (inputString[0] == "b") {
        inputString.remove(at: 0)
        parseString_C()
        return
    }
    
    if (inputString.isEmpty) {
        parserOK = false
    }
    else if (inputString[0] == "a") {
        inputString.remove(at: 0)
        parserOK = true
    }
}

func parseString_B() {
    parserOK = false
    print("B", terminator:"")
    
    if (inputString.isEmpty) {
        parserOK = true
    }
    else {
        parserOK = false
        if (inputString[0] == "c") {
            inputString.remove(at: 0)
            if (!inputString.isEmpty && inputString[0] == "c") {
                inputString.remove(at: 0)
                parseString_S()
                if (parserOK && !inputString.isEmpty && inputString[0] == "b") {
                    inputString.remove(at: 0)
                    if (!inputString.isEmpty && inputString[0] == "c") {
                        inputString.remove(at: 0)
                        parserOK = true
                    }
                    else {
                        parserOK = false
                    }
                }
                else {
                    parserOK = false
                }
            }
            else {
                parserOK = false
            }
        }
        else {
            parserOK = true
        }
    }
}

func parseString_C() {
    parserOK = false
    print("C", terminator:"")
    
    parseString_A()
    if (parserOK) {
        parseString_A()
    }
}

func checkDefinition() {
    print()
    
    if (parserOK && inputString.isEmpty) {
        print("DA")
    }
    else {
        print("NE")
    }
}

dataInput()

parseString_S()

checkDefinition()
